src/EVM/Fuzz.hs

```diff
@@ -0,0 +1,238 @@
+                                Just (Flat x) -> ConcreteBuf x
+                                Just (Comp _) -> internalError "No compressed allowed in fuzz"
+                                Nothing -> orig
+    go a = a
+
+substituteStores ::  Map (Expr 'EAddr) (Map W256 W256) -> Prop -> Prop
+substituteStores valMap p = mapProp go p
+  where
+    go :: Expr a -> Expr a
+    go (AbstractStore a) = case valMap !? a of
+                                  Just m -> ConcreteStore m
+                                  Nothing -> ConcreteStore mempty
+    go a = a
+
+-- Var extraction
+data CollectVars = CollectVars {vars :: Set.Set (Expr EWord)
+                               ,vals :: Set.Set W256
+                               }
+  deriving (Show)
+
+initVarsState :: CollectVars
+initVarsState = CollectVars {vars = Set.empty
+                            ,vals =  Set.empty
+                            }
+
+findVarProp :: Prop -> State CollectVars Prop
+findVarProp p = mapPropM go p
+  where
+    go :: forall a. Expr a -> State CollectVars (Expr a)
+    go = \case
+      e@(Var a) -> do
+        s :: CollectVars <- get
+        put CollectVars {vars=Set.insert (Var a) s.vars, vals = s.vals}
+        pure e
+      e@(Lit a) -> do
+        s :: CollectVars <- get
+        put $ s{vals=Set.insert a s.vals}
+        pure e
+      e -> pure e
+
+
+extractVars :: [Prop] -> CollectVars
+extractVars ps = execState (mapM_ findVarProp ps) initVarsState
+
+
+--- Buf extraction
+newtype CollectBufs = CollectBufs { bufs :: Set.Set (Expr Buf) }
+  deriving (Show)
+
+initBufsState :: CollectBufs
+initBufsState = CollectBufs { bufs = Set.empty }
+
+extractBufs :: [Prop] -> [Expr Buf]
+extractBufs ps = Set.toList bufs
+ where
+  CollectBufs bufs = execState (mapM_ findBufProp ps) initBufsState
+
+findBufProp :: Prop -> State CollectBufs Prop
+findBufProp p = mapPropM go p
+  where
+    go :: forall a. Expr a -> State CollectBufs (Expr a)
+    go = \case
+      e@(AbstractBuf a) -> do
+        s <- get
+        put $ s{bufs=Set.insert (AbstractBuf a) s.bufs}
+        pure e
+      e -> pure e
+
+--- Store extraction
+data CollectStorage = CollectStorage { addrs :: Set.Set (Expr EAddr)
+                                     , keys :: Set.Set W256
+                                     , values :: Set.Set W256
+                                     }
+  deriving (Show)
+
+initStorageState :: CollectStorage
+initStorageState = CollectStorage { addrs = Set.empty, keys = Set.empty, values = Set.fromList [0x0, 0x1, Expr.maxLit] }
+
+extractStorage :: [Prop] -> CollectStorage
+extractStorage ps = execState (mapM_ findStorageProp ps) initStorageState
+
+findStorageProp :: Prop -> State CollectStorage Prop
+findStorageProp p = mapPropM go p
+  where
+    go :: forall a. Expr a -> State CollectStorage (Expr a)
+    go = \case
+      e@(AbstractStore a) -> do
+        s <- get
+        put $  CollectStorage {addrs=Set.insert a s.addrs, keys=s.keys, values=s.values}
```

Comment:
```suggestion
        put $  s { addrs=Set.insert a s.addrs }
```

Reply:
See above. I just don't understand Haskell. Literally same exact error :cry:

Reply:
Thanks for helping with this! :heart: